cloud-controller-manager/do/loadbalancers.go

```diff
@@ -184,8 +197,19 @@ func newLoadBalancers(resources *resources, client *godo.Client, region string)
 // GetLoadBalancer returns the *v1.LoadBalancerStatus of service.
 //
 // GetLoadBalancer will not modify service.
-func (l *loadBalancers) GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (*v1.LoadBalancerStatus, bool, error) {
-	lb, err := l.retrieveAndAnnotateLoadBalancer(ctx, service)
+func (l *loadBalancers) GetLoadBalancer(ctx context.Context, clusterName string, service *v1.Service) (lbs *v1.LoadBalancerStatus, truthy bool, err error) {
+	patcher := newServicePatcher(service)
```

Comment:
@timoreimann @snormore here is an example of my solution for aggregating in-memory service modifications into a single PATCH call to the kubernetes api. The reason I chose to defer the Patch call is to reduce the chances of service changes happening that don't get captured before return. If you think that's too convoluted I'm fine identifying a more appropriate place to call `patcher.Patch`.

One thing that's worth noting is that in the defer methods as written if the function returns an error then we currently skip the patch. Do you think we should instead unconditionally always patch? I'm not certain myself which would be best.

Reply:
In general, I like the defer approach a lot. I'll comment around a few design details on more specific lines, but overall it seems like a good approach to ensure we include all patch-worthy annotations in a single invocation.

Yes, I think we should patch independently of whether we hit an error or not. That's what we did before, and the reason is this: even if our request to the LB API fails overall, the kind of patches we intend to do (LB ID / certificate updates) will help us avoid unnecessary follow-up API requests. For example, if we did not have the LB ID recorded in the Service yet and we only failed in a final LB update API request, persisting the LB ID annotation already would ensure that our API requests we executed in the controller-driven retry attempts could fetch the LB object efficiently in O(1) (vs. having to iterate through all LBs exhaustively in the worst case if the LB UUID was still lacking).

Similarly for the certificate ID where we could avoid reaching out to the Certificate API as soon as the certificate IDs matched.

The only thing we need to do then is to short-circuit our `Patch` call if the annotations match. That way, we don't even reach out to the API server if there's nothing worthwhile patching (and thus rule out a potential source of I/O erroring). Would appreciate if you could do the necessary adjustments. (We can copy [what upstream does](https://github.com/digitalocean/digitalocean-cloud-controller-manager/blob/67c685a86ca3b54897fa4f835c0e652844466a9a/vendor/k8s.io/kubernetes/pkg/controller/service/service_controller.go#L484-L486), which is to invoke `reflect.DeepCopy` on the annotations.)